agg-ast/schema_derivation/src/lib.rs

```diff
@@ -121,6 +121,82 @@ pub(crate) fn get_schema_for_path_mut(
     schema
 }
 
+/// Gets or creates a mutable reference to a specific field or document path in the schema. This
+/// should only be used in $match context where can can refine the schema. For example, we could
+/// think a field has type Any, or AnyOf([String, Document]), and a $match stage can only evaluate
+/// to true if that field is specifically a Document. In this case, we can refine that schema to
+/// Document, and this can recurse to any depth in the Schema. Note that this still returns an
+/// Option because if the field is known to have a Schema that cannot be a Document, we cannot
+/// create a path! This would mean that the aggregation pipeline in question will return no
+/// results, in fact, because the $match stage will never evaluate to true.
+pub(crate) fn get_or_create_schema_for_path_mut(
```

Comment:
my only q would be is this limited to $match? Couldn't we also use this for additive operations?

eg lets say a stage adds a field foo.bar with schema A, could we say, give me whatever the schema has in the path [foo, bar]. Then if it is Some(B), I can union A U B to get the new schema, otherwise B. Or am I spacing on a more ergonomic way to do this that is existing?

Reply:
As mentioned in standup, in most cases we would not want to use this, but there could be additive places where this is needed, so I'm updating the comment.